### ISSUE TITLE
A clearer explanation of data-weights with more details

### ISSUE THREAD

**Issue by leocnj** [EleutherAI/gpt-neox#992]
In [local_setup.yml](https://github.com/EleutherAI/gpt-neox/blob/main/configs/local_setup.yml), data-weights like `[1.0, 2.0]` is used.

However, the intended purpose of introducing such weights and the associated impacts to data loader are not quite clear.

When checking the code, it looks that these weights are normalized and passed to a C++ code to generate blending indices. 

https://github.com/EleutherAI/gpt-neox/blob/700219bbcd97df03b2cc14287ab9eedb9dc485ec/megatron/data/blendable_dataset.py#L39

Suppose we have two dataset, [A, B] with weights `[1., 2.]`, then the normalized weights will be `[0.33, 0.67]`. Does this mean that the blended indices will contain 33% from A while 67% from B?

**Comment by StellaAthena** [comment]
Yes that’s exactly right. It’s intended for people who have unmixed datasets (say, the components of the Pile or P3) and want to shuffle them together with sampling frequency other than the ratio of their sizes.

**Comment by leocnj** [comment]
@StellaAthena, thanks for your prompt reply and awesome works on gpt-neox!

I still have a question related to data-weights from my code reading.

In the Pile dataset, I saw the example of assigning weights to individual dataset
```
datasets = [
    # Academic
    (PubMedCentralDataset(), 2.  ),
    (ArXivDataset()        , 2.  ),
    (FreeLawDataset()      , 1.5 ),
    (USPTODataset()        , 2.  ),
```
Later, when computing `relative weight`, the following code is used
```
def mk_table(datasets, train_chars, print_latex=False):
    values = []

    total_weight = sum([x[1] * x[0].size() for x in datasets])

    for dataset, weight in datasets:
        size = dataset.size()
        relative_weight = size * weight / total_weight
        values.append([dataset.name(), size, '{:.2%}'.format(relative_weight), '{:.4f}'.format(train_chars / size * relative_weight), size * weight, humanbytes(size / dataset.num_docs(), 'KiB')])
```

It looks that when computing `relative_weight`, **both dataset weight `x[1]` and dataset size `x[0].size()` are used**.

However, in `BlendableDataset`, when computing `weights`, individual dataset's sizes are not considered.

```
class BlendableDataset(torch.utils.data.Dataset):
    def __init__(self, datasets, weights):
        self.datasets = datasets
        num_datasets = len(datasets)
        assert num_datasets == len(weights)

        self.size = 0
        for dataset in self.datasets:
            self.size += len(dataset)

        # Normalize weights.
        weights = np.array(weights, dtype=np.float64)
        sum_weights = np.sum(weights)
        assert sum_weights > 0.0
        weights /= sum_weights
```
Do you mind explaining why the different ways for computing weights are used in Pile vs. gpt-neox's data mixing code. Thanks

**Comment by StellaAthena** [comment]
The two pieces of code have different goals and were written two years apart by different people. The difference doesn’t really “matter” and each can be simulated by the other with a little work.
